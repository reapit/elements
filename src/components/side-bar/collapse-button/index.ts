export * from './collapse-button'
export * from './styles'
