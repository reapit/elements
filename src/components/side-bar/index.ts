export * from './side-bar'
export * from './styles'
