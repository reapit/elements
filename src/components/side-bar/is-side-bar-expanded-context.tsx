import { createContext, useContext, type Dispatch, type SetStateAction } from 'react'

interface IsSideBarExpandedContextValue {
  isExpanded: boolean
  setIsExpanded: Dispatch<SetStateAction<boolean>>
}
export const IsSideBarExpandedContext = createContext<IsSideBarExpandedContextValue | undefined>(undefined)

export const useIsSideBarExpandedContext = () => {
  const context = useContext(IsSideBarExpandedContext)
  if (!context) {
    throw new Error('useIsSideBarExpandedContext must be used within a SideBarExpandedContext Provider')
  }
  return context
}
