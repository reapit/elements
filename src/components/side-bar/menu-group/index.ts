export * from './styles'
export * from './menu-group'
export * from './menu-group-summary'
