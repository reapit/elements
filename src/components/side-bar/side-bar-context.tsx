import { createContext, useContext } from 'react'
import { useSideBar } from './use-side-bar'

const SideBarContext = createContext<useSideBar | null>(null)

const SideBarProvider = ({ children }) => {
  const sidebar = useSideBar()

  return <SideBarContext.Provider value={sidebar}>{children}</SideBarContext.Provider>
}

export { SideBarContext, SideBarProvider }

export const useSideBarContext = () => {
  const context = useContext(SideBarContext)
  if (!context) {
    throw new Error('useSideBarContext must be used within a SideBarProvider')
  }
  return context
}
