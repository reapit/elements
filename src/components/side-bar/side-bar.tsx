import type { FC, HTMLAttributes } from 'react'
import { ElSideBar } from './styles'

const SideBar: FC<HTMLAttributes<HTMLElement>> = ({ children, ...props }) => (
  <ElSideBar aria-label="Sidebar Navigations" {...props}>
    {children}
  </ElSideBar>
)

export { SideBar }
