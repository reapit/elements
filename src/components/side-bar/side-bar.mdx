
import { Canvas, Controls, Meta } from '@storybook/blocks'
import { RenderHtmlMarkup } from '../../storybook/render-html-markup'
import * as SideBarStories from './side-bar.stories'

<Meta of={SideBarStories} />

# Side Bar

SideBar component is part of the element's navigation, It is composed of several components to easily create a left-side-vertical navigation structure.

## Default

<Canvas of={SideBarStories.Default} />

<RenderHtmlMarkup of={SideBarStories.Default} />
