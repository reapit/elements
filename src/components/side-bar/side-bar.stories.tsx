import { MediaStateProvider } from '#src/hooks/use-media-query/index'
import { figmaDesignUrls } from '#src/storybook/figma/index'
import type { Meta, StoryObj } from '@storybook/react'
import { Icon } from '../icon'
import { SideBarMenuGroup, SideBarMenuGroupItem } from '../side-bar-menu-group'
import {
  elSideBarMenuItemAnchor,
  ElSideBarMenuItemIcon,
  ElSideBarMenuItemText,
  SideBarMenuItem,
} from '../side-bar-menu-item'
import { useIsSideBarExpandedContext } from './is-side-bar-expanded-context'
import { SideBar } from './side-bar'

export default {
  title: 'Components/Side Bar',
  component: SideBar,
} as Meta<typeof SideBar>

type Story = StoryObj<typeof SideBar>

const Customicon = () => (
  <svg width={24} height={24} viewBox="0 0 24 24" fill="none" xmlns="http://www.w3.org/2000/svg">
    <circle cx="12" cy="8" r="4" stroke="currentColor" strokeWidth="2" />
    <path d="M4 20c0-4 4-6 8-6s8 2 8 6" stroke="currentColor" strokeWidth="2" strokeLinecap="round" />
  </svg>
)

export const Default: Story = {
  render: () => {
    const CustomLink = () => {
      const { isExpanded } = useIsSideBarExpandedContext()

      return (
        <a href="#" className={elSideBarMenuItemAnchor}>
          <ElSideBarMenuItemIcon>
            <Icon icon="property" />
          </ElSideBarMenuItemIcon>
          {isExpanded && <ElSideBarMenuItemText>Custom Link</ElSideBarMenuItemText>}
        </a>
      )
    }
    return (
      <MediaStateProvider>
        <div
          style={{
            display: 'flex',
            flexDirection: 'row',
            height: '100%',
            overflow: 'hidden',
          }}
        >
          <style>
            {`
            #storybook-root {
              height: 100%;
            }
          `}
          </style>

          <SideBar>
            <SideBar.MenuList>
              <SideBarMenuItem isActive icon={<Icon icon="property" />} href="#">
                SideBar Item (active)
              </SideBarMenuItem>
              <SideBarMenuGroup isActive label="Menu Group 1" icon={<Icon icon="property" />}>
                <SideBarMenuGroupItem isActive href="#">
                  Sub Menu Item 1
                </SideBarMenuGroupItem>
                <SideBarMenuGroupItem href="#">Sub Menu Item 2</SideBarMenuGroupItem>
              </SideBarMenuGroup>

              <SideBarMenuGroup label="Menu Group 2" icon={<Icon icon="property" />}>
                <SideBarMenuGroupItem href="#">Sub Menu Item 3</SideBarMenuGroupItem>
                <SideBarMenuGroupItem href="#">Sub Menu Item 4</SideBarMenuGroupItem>
              </SideBarMenuGroup>

              <li>
                <CustomLink />
              </li>
              <SideBarMenuItem icon={<Customicon />} href="#">
                External Icon
              </SideBarMenuItem>

              {[1, 2, 3, 4, 5, 6, 7, 8, 9, 10].map((i) => {
                return (
                  <SideBarMenuItem key={i} icon={<Icon icon="property" />} href="/#">
                    SideBar Item
                  </SideBarMenuItem>
                )
              })}
            </SideBar.MenuList>
            <SideBar.CollapseButon />
          </SideBar>
          <main
            style={{
              width: '100%',
              overflow: 'auto',
            }}
          >
            <div
              style={{
                background: '#ffdaf5',
                paddingTop: '13rem',
                paddingLeft: '3rem',
                fontSize: '1.25rem',
                height: '115vh',
              }}
            >
              Placeholder main content
            </div>
          </main>
        </div>
      </MediaStateProvider>
    )
  },
  parameters: {
    design: {
      type: 'figma',
      url: figmaDesignUrls.sideBar,
      allowFullscreen: true,
    },
  },
}
