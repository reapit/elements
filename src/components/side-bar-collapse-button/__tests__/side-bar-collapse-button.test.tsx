import { useIsSideBarExpandedContext } from '#src/components/side-bar/is-side-bar-expanded-context'
import { render, screen } from '@testing-library/react'
import type { Mock } from 'vitest'
import { SideBarCollapseButton } from '..'

vi.mock('../../side-bar-collapse-button/icons/collapse.svg?react', () => ({
  default: () => <span data-testid="collapse-icon" />,
}))

vi.mock('#src/components/side-bar/is-side-bar-expanded-context', () => ({
  useIsSideBarExpandedContext: vi.fn(),
}))

const mockuseIsSideBarExpandedContext = useIsSideBarExpandedContext as Mock

describe('SideBar', () => {
  beforeAll(() => {
    mockuseIsSideBarExpandedContext.mockReturnValue({ isExpanded: true, setIsExpanded: vi.fn() })
  })

  it('should match snapshot with expanded state', () => {
    const { asFragment } = render(<SideBarCollapseButton />)

    expect(screen.getByRole('button')).toHaveAttribute('aria-expanded', 'true')
    expect(asFragment()).toMatchSnapshot()
  })

  it('should handle click properly and match snapshot with collapsed state', () => {
    const mockOnClick = vi.fn()
    mockuseIsSideBarExpandedContext.mockReturnValue({ isExpanded: false, setIsExpanded: mockOnClick })
    const { asFragment } = render(<SideBarCollapseButton />)

    expect(screen.getByRole('button')).toHaveAttribute('aria-expanded', 'false')
    screen.getByRole('button').click()

    expect(mockOnClick).toHaveBeenCalledTimes(1)
    expect(asFragment()).toMatchSnapshot()
  })
})
