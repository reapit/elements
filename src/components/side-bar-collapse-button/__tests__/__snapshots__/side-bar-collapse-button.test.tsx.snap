// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`SideBar > should handle click properly and match snapshot with collapsed state 1`] = `
<DocumentFragment>
  <div
    class="mocked-styled-2 el-side-bar-collapse-button-container"
    data-is-expanded="false"
  >
    <button
      aria-expanded="false"
      class="mocked-styled-3 el-side-bar-collapse-button"
    >
      <span
        data-testid="collapse-icon"
      />
    </button>
  </div>
</DocumentFragment>
`;

exports[`SideBar > should match snapshot with expanded state 1`] = `
<DocumentFragment>
  <div
    class="mocked-styled-2 el-side-bar-collapse-button-container"
    data-is-expanded="true"
  >
    <button
      aria-expanded="true"
      class="mocked-styled-3 el-side-bar-collapse-button"
    >
      <span
        data-testid="collapse-icon"
      />
      <span
        class="mocked-styled-5 el-side-bar-collapse-text"
      >
        Collapse
      </span>
    </button>
  </div>
</DocumentFragment>
`;
